engine/src/main/java/org/terasology/physics/engine/VoxelFluidWorldSystem.java

```diff
@@ -0,0 +1,157 @@
+import org.terasology.entitySystem.systems.BaseComponentSystem;
+import org.terasology.entitySystem.systems.RegisterSystem;
+import org.terasology.math.geom.Vector3i;
+import org.terasology.physics.StandardCollisionGroup;
+import org.terasology.physics.bullet.BulletPhysics;
+import org.terasology.physics.bullet.shapes.BulletCollisionShape;
+import org.terasology.registry.In;
+import org.terasology.world.OnChangedBlock;
+import org.terasology.world.WorldComponent;
+import org.terasology.world.WorldProvider;
+import org.terasology.world.block.Block;
+import org.terasology.world.block.BlockComponent;
+import org.terasology.world.chunks.Chunk;
+import org.terasology.world.chunks.ChunkConstants;
+import org.terasology.world.chunks.ChunkProvider;
+import org.terasology.world.chunks.event.BeforeChunkUnload;
+import org.terasology.world.chunks.event.OnChunkLoaded;
+
+import java.nio.ByteBuffer;
+import java.nio.ByteOrder;
+
+import static org.terasology.physics.bullet.BulletPhysics.AABB_SIZE;
+
+/**
+ * Manages voxel shape for fluid and updates collision state between Bullet and Terasology
+ */
+@RegisterSystem
+public class VoxelFluidWorldSystem extends BaseComponentSystem {
+    @In
+    private PhysicsEngine physics;
+    @In
+    private EntityManager entityManager;
+    @In
+    private WorldProvider worldProvider;
+    @In
+    private ChunkProvider chunkProvider;
+
+    private final TShortHashSet registered = new TShortHashSet();
+
+    private btRigidBody.btRigidBodyConstructionInfo blockConsInf;
+    private btVoxelShape worldShape;
+    private VoxelCollisionAlgorithmWrapper wrapper;
+    private btRigidBody rigidBody;
+
+    @Override
+    public void initialise() {
+        if (physics instanceof BulletPhysics) {
+            btDiscreteDynamicsWorld discreteDynamicsWorld = ((BulletPhysics) physics).getDiscreteDynamicsWorld();
+
+            wrapper = new VoxelCollisionAlgorithmWrapper(ChunkConstants.SIZE_X, ChunkConstants.SIZE_Y,
+                ChunkConstants.SIZE_Z);
+            worldShape = new btVoxelShape(wrapper, new Vector3f(-AABB_SIZE, -AABB_SIZE, -AABB_SIZE),
+                new Vector3f(AABB_SIZE, AABB_SIZE, AABB_SIZE));
+
+            Matrix4f matrix4f = new Matrix4f();
+            btDefaultMotionState blockMotionState = new btDefaultMotionState(matrix4f);
+
+            blockConsInf = new btRigidBody.btRigidBodyConstructionInfo(0, blockMotionState, worldShape, new Vector3f());
+            rigidBody = new btRigidBody(blockConsInf);
+            rigidBody.setCollisionFlags(btCollisionObject.CollisionFlags.CF_STATIC_OBJECT | rigidBody.getCollisionFlags()); // voxel world is added to static collision flag
+            short mask = (short) StandardCollisionGroup.LIQUID.getFlag(); // interacts with anything but static and
+            // liquid
+            discreteDynamicsWorld.addRigidBody(rigidBody, physics.combineGroups(StandardCollisionGroup.WORLD), mask); // adds rigid body to world
```

Comment:
Could you describe what exactly the mask and the rigid body's collision group are for? As I understand it, the collision group specified on line 82 is the group that this mesh is itself a part of, so that if a raytrace is requested, this group determines whether this mesh is included, therefore it should be LIQUID, not WORLD, and the mask determines what other groups this collides with, which should probably just be nothing (0), because liquids don't actually collide. I expect it would still be possible to make something collide with this by including LIQUID in _its_ collision mask (e.g. potentially a duck or a player wearing boots of water-walking or something).